libcudacxx/include/cuda/buffer

```diff
@@ -0,0 +1,187 @@
+
+#if _CCCL_STD_VER >= 2014 && !defined(_CCCL_COMPILER_MSVC_2017) \
+  && defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+
+//! @file The cuda::mr::uninitialized_buffer class provides a typed buffer allocated from a given memory resource.
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_MR
+
+//! @rst
+//! .. _libcudacxx-extended-api-memory-resources-buffer:
+//! The `cuda::mr::uninitialized_buffer` memory storage
+//! -----------------------------------------------
+//!
+//! ``cuda::mr::uninitialized_buffer`` provides a typed buffer allocated from a given `memory resource
+//! <libcudacxx-extended-api-memory-resources-resource>`. It handles alignment and release of the allocation.
+//! The memory is uninitialized, so that a user needs to ensure elements are properly constructed.
+//! In addition to being type safe, ``cuda::mr::uninitialized_buffer`` also takes a set of `properties
+//! <libcudacxx-extended-api-memory-resources-properties>` to ensure that e.g. execution space constraints are checked
+//! at compile time. However, we can only forward stateless propertie. If a user wants to a a stateful one, then they
+//! need to implement ``get_property(const device_buffer&, Property)``.
+//!
+//! .. note::
+//!
+//!    ``cuda::mr::uninitialized_buffer`` stores a reference to the provided memory `memory resource
+//!    <libcudacxx-extended-api-memory-resources-resource>`. It is the users resposibility to ensure the lifetime of the
+//!    resource exceeds the lifetime of the buffer.
+//!
+//! @endrst
+template <class _Tp, class... _Properties>
+class uninitialized_buffer
+{
+private:
+  _CUDA_VMR::resource_ref<_Properties...> __mr_;
+  size_t __count_ = 0;
+  void* __buf_    = nullptr;
+
+  //! @brief Determines the allocation size given the alignment and size of @p _Tp
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY static constexpr size_t
+  __get_allocation_size(const size_t __count) noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    return (__count * sizeof(_Tp) + (__alignment - 1)) & ~(__alignment - 1);
+  }
+
+  //! @brief Determines the properly aligned start of the buffer given the alignment and size of @p _Tp
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr _Tp* __get_data() const noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    size_t __space               = __get_allocation_size(__count_);
+    void* __ptr                  = __buf_;
+    return reinterpret_cast<_Tp*>(_CUDA_VSTD::align(__alignment, __count_ * sizeof(_Tp), __ptr, __space));
+  }
+
+public:
+  //! @brief Constructs a ``cuda::uninitialized_buffer`` from a memory resource and a number of elements
+  //! @param __mr The memory resource to allocate the buffer with.
+  //! @param __count The desired size of the buffer.
+  uninitialized_buffer(_CUDA_VMR::resource_ref<_Properties...> __mr, const size_t __count)
+      : __mr_(__mr)
+      , __count_(__count)
+      , __buf_(__mr_.allocate(__get_allocation_size(__count_)))
```

Comment:
I was wondering about an optimization: what if we stored `__buf_(__get_data(__mr_.allocate(__get_allocation_size(__count_))))` instead? That would take out some cost on each access, and only complicate deallocation, which would need to reconstruct the original address obtained from the allocator. Worst case cost is an extra pointer to the allocation in addition to the start of the aligned data. But if the alignment is a low number, the offset between the pointers is likely small enough to fit in some unused bits of `__buf_` of `__count`.

But this can all be done later, if aligning during data access becomes an issue.

Reply:
Yeah I thought about this to inititally, but I found that the reconstruction of the initial information was too much for this